src/python/WMCore/WMSpec/Steps/Executors/StageOut.py

```diff
@@ -129,102 +129,98 @@ def execute(self, emulator = None):
+            # with the output module as key (straight to merge or not
+            # can be output module specific, so we need to know...
+            fileDict = stepReport.getAllFileRefsFromStep(step = step)
+            for outputName in fileDict:
+                for fileName in fileDict[outputName]:
+
+                    if not hasattr(fileName, 'lfn') and hasattr(fileName, 'pfn'):
+                        # Then we're truly hosed on this file; ignore it
+                        msg = "Not a file: %s" % fileName
+                        logging.error(msg)
                         continue
-                    if fileName.size >= self.step.output.minMergeSize:
-                        # Then this goes direct to merge
+
+                    # Figuring out if we should do straight to merge
+                    # First check if we should do straight to merge at all,
+                    # then check if we are over the size threshold, then check
+                    # if we are over the event threshold
+                    straightToMerge = False
+                    if hasattr(self.step.output, 'minMergeSize'):
+                        if outputName not in getattr(self.step.output, 'forceUnmergedOutputs', []):
+                            if not getattr(fileName, 'merged', False):
+                                if hasattr(fileName, 'size'):
+                                    if fileName.size >= self.step.output.minMergeSize:
+                                        straightToMerge = True
+                                if hasattr(self.step.output, 'maxMergeEvents'):
+                                    if hasattr(fileName, 'events'):
+                                        if fileName.events >= self.step.output.maxMergeEvents:
+                                            straightToMerge = True
+
+
+                    if straightToMerge:
+
                         try:
-                            fileName = self.handleLFNForMerge(mergefile = fileName, step = step)
+                            fileName = self.handleLFNForMerge(mergefile = fileName,
+                                                              step = step,
+                                                              outputName = outputName)
                         except Exception as ex:
-                            logging.error("Encountered error while handling LFN for merge due to size.\n")
+                            logging.error("Encountered error while handling LFN for merge.\n")
                             logging.error(str(ex))
                             logging.debug(fileName)
-                            logging.debug("minMergeSize: %s" % self.step.output.minMergeSize)
+                            logging.debug("minMergeSize: %s" % getattr(self.step.output, 'minMergeSize', None))
+                            logging.debug("maxMergeEvents: %s" % getattr(self.step.output, 'maxMergeEvents', None))
```

Comment:
Since we lack of good logging/report on the stageout land, it's likely a good idea to move these two min/maxMerge debug logging to the beginning of the exception block AND make it an info log level.